### ISSUE TITLE
[2008.10546] SDE-Net: Equipping Deep Neural Networks with Uncertainty Estimates

### ISSUE THREAD

**Issue by riron1206** [riron1206/read_paper#77]
## 概要
https://arxiv.org/abs/2008.10546
SDE-Net
Neural ordinaryで不確実性を推定する手法
ResNetを常微分方程式（ODE: Ordinary Differential Equations）として考えたNeural ordinary differential equations( https://arxiv.org/abs/1806.07366. この論文はNeurIPS2018のbest paperとったらしい )に確率的な手法を追加して不確実性を推定できるようにした
aleatoric uncertainty（データに内在された不確実性）とepistemic uncertainty（モデルがわからない不確実性）を区別できる

【特徴】
- Deepアンサンブルと比較して効果あること示してる（Deepアンサンブルは2種の不確実性分離できない）
- BNNのようにモデルの事前分布を指定し、事後分布を推測する必要がないため、実装は効率的で簡単
- 分類タスクと回帰タスクの両方に適用できます
-  OOD（out of distribution）detection、Misclassification Detection、Adversarial Sample Detection、Active Learningの両方で良好なパフォーマンスを見せる

【ネットワーク】
Drift Net
- 良い予測精度を学習することに目標を置く
- Aleatoric uncertaintyを測定することを目的とする
- Regression問題の場合meanとvarianceを出力してNLLに学習させる（Deep Ensembleと同じ方法）
- Classificationはmeanを出力してcross entropy errorに学習させる。学習をするときbinary cross entropy errorを使用してfake（OOD）とtrue（ID）をよく区別するように学習させる

Diffusion Net
- epistemice uncertaintyを求めることの目標を置く

【不確実性の計算方法】
- Aleatoric uncertainty：Regressionの場合FC layerを通じて出力されるvarianceのsamplingされたmean値を使用して求めることができる。Classification問題はentropyから入手することができる。 
- Epistemic uncertainty：x_Tのvariance

【欠点】
**SDE-Netはトレーニングプロセス中にOODサンプルを必要とするため、デフォルトでは、ガウスノイズによってトレーニングデータを疑似OODデータとして摂動させます**

## 手法
OED Net のResNet出力は以下
![image](https://user-images.githubusercontent.com/22429441/112603329-3311c180-8e58-11eb-8753-a01bdfa6bd77.png)

SDE NetのResNet出力は以下。確率的な要素としてブラウン運動の項が追加される
f：drift net、g：diffusion net
![image](https://user-images.githubusercontent.com/22429441/112603292-255c3c00-8e58-11eb-9b2d-8c5469076104.png)

fは予測をよくこなすことが目標であり、 gは不確実性を知ることが目的

SDE-Netの構成要素
![image](https://user-images.githubusercontent.com/22429441/112603224-0d84b800-8e58-11eb-9676-bad45b9b65a9.png)
学習と同じ分布のデータの場合、drift net fによって支配され、優れた予測精度を実現します
学習の分布外データの場合、diffusion net gによって支配されます
上の図のMNISTの画像は分布内のデータだからfが支配的になり、出力はばらつき少ない。犬の画像は分布外のデータだからgが支配的になり、出力はばらつき大きくなって=epistemic uncertaintyでかい

## 結果
MNIST、SVHNなどでOut of distribution detection、Misclassification Detection、Adversarial Sample Detection、Active Learningについて
MC-dropout、DeepEnsemble、Prior Network（PN）、Bayes By Backpropagation（BBP）、preconditioned Stochastic gradient Langevin dynamics（p-SGLD）
の手法と比較

- OODのaccuracyはDeepEnsembleモデルが一番良い。TNR、AUROC、AUPRはSDE-Netが一番
- Misclassification detection結果はDeepEnsembleがいい。しかし、SDE-Netとあまり差がない
- active learningはSDE-NetとMC-dropoutが良い結果を得た。DeepEnsemble方法は良くない結果を出した

## コメント

公式github:
- https://github.com/Lingkai-Kong/SDE-Net  # pytorch

github
- https://github.com/makora9143/deterministic-variational-inference-pytorch  # pytorch

参考サイト
- https://simpling.tistory.com/31
- https://www.programmersought.com/article/77686143253/
- https://papertalk.org/papertalks/5819  # 著者のスライド

## メモ
- 微分方程式: 未知関数の導関数を含む方程式
- 常微分方程式: 独立変数が1つだけの微分方程式( dx/dt = ax(t) )。常微分方程式の解析解を求めるのは困難なのである時刻tにおける値を数値解として求める=ODE Solver( オイラー法、ルンゲクッタ法 etc )
https://www.slideshare.net/DeepLearningJP2016/dlneural-ordinary-differential-equations

決定論的記述ができない現象を考察するとき，ランダムな揺らぎを数学的に記述するのが確率微分方程式
確率微分方程式は，大まかに言えば時間とともに偶然に変化する要因を伴った常微分方程式
https://www.shinshu-u.ac.jp/faculty/science/quest/research/post-4.php

ブラウン運動では粒子はランダムに動く。
そして試行をくり返せばその統計集団は正規分布に従う。
そして分布の標準偏差は時間tの時には √t に比例する。
これがブラウン運動の基本的な性質
http://saiyu.cocolog-nifty.com/zug/2006/12/post_d4a7.html

---------------------------------------------------------------------------------------------------

- 分布外のデータ検出＝モデルの不確実性高いデータ検出。分布外のデータは学習不足であるので
- 画像に分類クラスの対象がピンボケで写ってるなど入力にノイズが多いデータ検出＝データの不確実性高いの検出
- 分類間違う予測難しいハードサンプル検出（誤分類検出）＝データの不確実性高いの検出。学習不足の可能性もあるのでモデルの不確実性高いのも検出する必要あり
- 能動学習＝学習効率上げるためにモデルの不確実性高くてデータの不確実性低いデータを学習に加える
